File: src/json/minimal_dual_keys.json.rb
#!/usr/bin/env ruby

PARAMETERS = {
  :to_if_alone_timeout_milliseconds => 300,
  :to_delayed_action_delay_milliseconds => 0,
  :to_if_held_down_threshold_milliseconds => 0,
  :simultaneous_threshold_milliseconds => 300,
}.freeze

require 'json'
require_relative '../lib/karabiner.rb'

def main
  puts JSON.pretty_generate(
    "title" => "Minimal dual keys",
    "maintainers" => ["yeonsh"],
    "rules" => [
      {
        "description" => "Minimal dual keys",
        "manipulators" => [
          generate_dual_key_rule("left_command", "delete_or_backspace", "left_command"),
          generate_dual_key_rule("right_command", "return_or_enter", "right_command"),
          generate_dual_key_rule("left_shift", "return_or_enter", "left_shift"),
        ],
      },
    ],
  )
end

def generate_dual_key_rule(input, alone, held_down)
  {
    "type" => "basic",
    "from" => {
      "key_code" => input,
      "modifiers" => { "optional" => ["any"] },
    },
    "to_if_alone" => [
      {
        "key_code" => alone,
      },
    ],
    "to_if_held_down" => [
      {
        "key_code" => held_down,
      },
    ],
    'parameters' => {
      'basic.to_if_alone_timeout_milliseconds' => PARAMETERS[:to_if_alone_timeout_milliseconds],
      'basic.to_if_held_down_threshold_milliseconds' => PARAMETERS[:to_if_held_down_threshold_milliseconds],
    },
  }
end

def generate_single_key_rule(input, output)
  {
    "type" => "basic",
    "from" => {
      "key_code" => input,
      "modifiers" => { "optional" => ["any"] },
    },
    "to" => [
      {
        "key_code" => output,
      },
    ],
  }
end

main()

File: src/json/minimal_dual_keys.json.js
// JavaScript should be written in ECMAScript 5.1.

const parameters = {
  to_if_alone_timeout_milliseconds: 300,
  to_delayed_action_delay_milliseconds: 0,
  to_if_held_down_threshold_milliseconds: 0,
  simultaneous_threshold_milliseconds: 300,
}

function main() {
  console.log(
    JSON.stringify(
      {
        title: 'Minimal dual keys',
        maintainers: ['yeonsh'],
        rules: [
          {
            description: 'Minimal dual keys',
            manipulators: [
              generateDualKeyRule('left_command', 'delete_or_backspace', 'left_command'),
              generateDualKeyRule('right_command', 'return_or_enter', 'right_command'),
              generateDualKeyRule('left_shift', 'return_or_enter', 'left_shift'),
            ],
          },
        ],
      },
      null,
      '  '
    )
  )
}

function generateDualKeyRule(input, alone, heldDown) {
  return {
    type: 'basic',
    from: {
      key_code: input,
      modifiers: { optional: ['any'] },
    },
    to_if_alone: [
      {
        key_code: alone,
      },
    ],
    to_if_held_down: [
      {
        key_code: heldDown,
      },
    ],
    parameters: {
      'basic.to_if_alone_timeout_milliseconds': parameters.to_if_alone_timeout_milliseconds,
      'basic.to_if_held_down_threshold_milliseconds': parameters.to_if_held_down_threshold_milliseconds,
    },
  }
}

main()
